README
flvtool++ is a tool for hinting and manipulating the metadata of FLV files. It was originally created for Facebook's Video project (http://facebook.com/video/) for fast video hinting. It is loosely based on the Ruby FLVTool2, but is written in C++ for performance reasons.

To get the latest version, go to: http://developers.facebook.com/opensource.php

Requirements:
  * SCons (http://www.scons.org/)
  * Boost >= 1.33.1 (http://www.boost.org/)

Compiling:
  To compile, run "scons" wherever you untarred the code.

Running:
  Run flvtool++ with no arguments to learn about its usage.

Notes:
  This code is provided with absolutely no support from us, but patches and qualified bug reports are welcomed at opensource@facebook.com.


CHANGELOG
flvtool++ Changelog
------------------------------------------------------
* 1.0 (September 7th, 2007) - Initial release
* 1.1 (March 17th, 2008) - Added -strip, better handling of flags (always rebuilt when writing a new file), fixed some compile errors reported by users
* 1.1.1 (September 3rd, 2008) - Add support for the TimestampExtended field in the FLV tag header, which allows properly calculating the duration of videos >4.5 hours long.
* 1.1.2 (September 8th, 2008) - Detect and correct invalid TimestampExtended field contents that would cause timestamps to appear to wrap around; also fixes duration calculation for those files.
* 1.1.3 (October 7, 2008) - Better bounds checking on metadata decoding to prevent crashes on files with malformed onMetaData tags.
* 1.1.4 (October 8, 2008) - Handle unknown tag types without an assertion failure.
* 1.1.5 (October 16, 2008) - Correctly propogate errors encountered while opening output files.
* 1.1.6 (November 4, 2008) - Allow flvtool++ to build on Mac OS X-based systems.
* 1.1.7 (February 5, 2009) - Applied patches contributed by Thomas Loo <info@saltstorm.net>, Saltstorm Software Labs:
        * Copy metadata packets to the output stream by default
        * Added -nometapackets option
        * Add support for AMF types UNSUPPORTED, NULL, and UNDEFINED
        * Ignore zero-size audio packets generated by Adobe FMS Netstream::record
* 1.2.0 (June 12, 2009) - Add description strings for more audio and video format types, add support for reading dimensions from H.264 video streams.
* 1.2.1 (June 18, 2009) - Handle truncated streams better. Add some missing includes to <fout.h>.


fout.h
/*
 * fout.h
 *
 * Copyright (c) 2007-2009 Dan Weatherford and Facebook, inc.
 * All rights reserved.
 */

#pragma once
#define BUFFER_SIZE 32768
#include <string.h> // for strerror
#include <errno.h>
#include <stdint.h>
#include <cstdio>

class fout {
public:
  fout() : fp(NULL), buffer_used(0) {}
  fout(const char* fn) : fp(NULL), buffer_used(0) { this->open(fn); }
  ~fout() { close(); }

  void open(const char* fn) {
    if (fp) this->close();

    fp = fopen(fn, "wb");
    if (fp == NULL) {
      char errbuf[256];
      snprintf(errbuf, 255, "Error opening output file \"%s\": %s", fn, strerror(errno));
      errbuf[255] = '\0';
      throw std::runtime_error(errbuf);
    }
  }

  operator bool() const {
    return (fp != NULL);
  }

  void flush() {
    if (buffer_used) fwrite(buffer, buffer_used, 1, fp);
    buffer_used = 0;
  }

  void close() {
    if (fp) {
      this->flush();
      fclose(fp);
      fp = NULL;
    }
  }

  void write(const char* dat, size_t len) {
    if ((len + buffer_used) > BUFFER_SIZE) {
      this->flush();
    }
    if (len > BUFFER_SIZE) fwrite(dat, len, 1, fp);
    else {
      memcpy(buffer + buffer_used, dat, len);
      buffer_used += len;
    }
  }

  void write_u24_be(uint32_t d) {
    putc((d >> 16) & 0xff);
    putc((d >> 8) & 0xff);
    putc(d & 0xff);
  }

  void putc(char c) {
    if (buffer_used == BUFFER_SIZE) this->flush();
    buffer[buffer_used++] = c;
  }

  uint64_t tell() {
    return (ftello(fp) + ((uint64_t)buffer_used));
  }

  void seek(uint64_t offset, int whence = SEEK_SET) {
    this->flush();
    fseek(fp, offset, whence);
  }

  template <typename T> inline void write(T d) {
    this->write((const char*)&d, sizeof(T));
  }

  template <class Str> inline void write_string(const Str& s) {
    uint32_t sl = s.size();
    this->write((const char*)&sl, sizeof(uint32_t));
    if (sl) this->write(s.data(), sl);
  }

protected:
  FILE* fp;
  uint32_t buffer_used;
  char buffer[BUFFER_SIZE];
private:
  fout(const fout& _r); // noncopyable
  fout& operator=(const fout& _r); // nonassignable
} ;



mmfile.h
/* 
 * mmfile.h
 *
 * Copyright (c) 2007-2009 Dan Weatherford and Facebook, inc.
 * All rights reserved.
 */

#pragma once

#include <sys/stat.h>
#include <sys/mman.h>
#include <fcntl.h>
#include <stdexcept>

class mmfile {
public:
  mmfile() : fd(-1) {} 
  mmfile(char* fn) {
    fd = open(fn, O_RDONLY);
    if (fd == -1) throw std::runtime_error(string("mmfile: unable to open file ") + string(fn));
    struct stat statbuf;
    fstat(fd, &statbuf);
    flen = statbuf.st_size;
    fbase = (char*) mmap(NULL, flen, PROT_READ, MAP_SHARED, fd, 0);
  }

  ~mmfile() {
    this->close();
  }

  void close() {
    if (fd != -1) {
      munmap(fbase, flen);
      ::close(fd); // yep, namespacing. call the stdlib close()
      fd = -1;
    }
  }

  char* fbase;
  size_t flen;
  int fd;
private:
  mmfile(const mmfile& right); // noncopyable
  mmfile& operator=(const mmfile& right); // nonassignable
} ;



AMFData.cpp
/*
 * AMFData.cpp
 *
 * Copyright (c) 2007-2008 Dan Weatherford and Facebook, inc.
 * All rights reserved.
 */

#include "AMFData.h"

shared_ptr<AMFData> AMFData::construct(serialized_buffer& buf) {
  char typeID = buf.get_u8();
  switch ((typeID & 0xff)) {
    case AMF_TYPE_DOUBLE:
      return shared_ptr<AMFData>(new AMFDouble(buf));
    case AMF_TYPE_BOOLEAN:
      return shared_ptr<AMFData>(new AMFBoolean(buf));
    case AMF_TYPE_STRING:
      return shared_ptr<AMFData>(new AMFString(buf));
    case AMF_TYPE_OBJECT:
      return shared_ptr<AMFData>(new AMFObject(buf));
    /* http://osflash.org/documentation/amf/astypes#x06null */
    case AMF_TYPE_NULL:
      return shared_ptr<AMFData>(new AMFNull(buf));
    /* http://osflash.org/documentation/amf/astypes#x06undefined */
    case AMF_TYPE_UNDEFINED:
      return shared_ptr<AMFData>(new AMFUndefined(buf));
    case AMF_TYPE_MIXED_ARRAY:
      return shared_ptr<AMFData>(new AMFMixedArray(buf));
    case AMF_TYPE_ARRAY:
      return shared_ptr<AMFData>(new AMFArray(buf));
    case AMF_TYPE_DATE:
      return shared_ptr<AMFData>(new AMFDate(buf));
    /* http://osflash.org/documentation/amf/astypes#x06unsupported */
    case AMF_TYPE_UNSUPPORTED:
      return shared_ptr<AMFData>(new AMFUnsupported(buf));
  }
  /// default:
  char errbuf[64];
  sprintf(errbuf, "AMFData::construct: unknown typeID 0x%02x\n", (typeID & 0xff));
  throw std::runtime_error(errbuf);
  return shared_ptr<AMFData>(new AMFData());
}




bitstream.h
/*
 * bitstream.h
 *
 * Copyright (c) 2007-2009 Dan Weatherford and Facebook, inc.
 * All rights reserved.
 */

#pragma once
#include "serialized_buffer.h"
#include <assert.h>

class bitstream {
public:
  bitstream(serialized_buffer* _buffer) : buffer(_buffer), current_byte(0), last_used_bit(0) {}

  // I'm sure this is really slow, but this isn't really in a critical path...
  uint32_t get_bits(uint8_t nbits) {
    uint32_t result = 0;
    assert(nbits <= 32);
    while (nbits) {
      result = result << 1 | get_bit();
      --nbits;
    }
    return result;
  }

  uint8_t get_bit() {
    if (! last_used_bit) refill();
    --last_used_bit;
    return ((current_byte & (1 << last_used_bit)) >> last_used_bit) & 1;
  }

  uint32_t get_golomb_ue() {
    uint32_t leading_zeros = 0;
    while (! get_bit()) ++leading_zeros;
    return ((1 << leading_zeros) | get_bits(leading_zeros)) - 1;
  }

  int32_t get_golomb_se() {
    uint32_t ue = get_golomb_ue();
    if (! ue) return 0;
    else if (ue & 1) return (ue >> 1);
    else return (0 - (ue >> 1));
  }

protected:
  void refill() {
    assert(last_used_bit == 0);
    current_byte = buffer->get_u8();
    last_used_bit = 8;
  }

  serialized_buffer* buffer;
  uint8_t current_byte;
  uint8_t last_used_bit;

private:
  bitstream& operator=(const bitstream&);
  bitstream(const bitstream&);
};



serialized_buffer.h
/*
 * serialized_buffer.h
 *
 * Copyright (c) 2007-2009 Dan Weatherford and Facebook, inc.
 * All rights reserved.
 */

#pragma once
#include <stdexcept>

#ifdef __APPLE__ // Mac byteswap defs

#include <architecture/byte_order.h>
#define BE16(x) OSSwapBigToHostInt16(x)
#define BE32(x) OSSwapBigToHostInt32(x)
#define BE64(x) OSSwapBigToHostInt64(x)
#define LE16(x) OSSwapLittleToHostInt16(x)
#define LE32(x) OSSwapLittleToHostInt32(x)
#define LE64(x) OSSwapLittleToHostInt64(x)

#else // Linux byteswap defs

#include <byteswap.h>
#include <endian.h>

#if __BYTE_ORDER == __LITTLE_ENDIAN
  #define BE16(x) __bswap_16(x)
  #define BE32(x) __bswap_32(x)
  #define BE64(x) __bswap_64(x)
  #define LE16(x) (x)
  #define LE32(x) (x)
  #define LE64(x) (x)
#else
  #define BE16(x) (x)
  #define BE32(x) (x)
  #define BE64(x) (x)
  #define LE16(x) __bswap_16(x)
  #define LE32(x) __bswap_32(x)
  #define LE64(x) __bswap_64(x)
#endif

#endif

class end_of_buffer : public std::exception {
public:
  end_of_buffer(size_t needed, size_t available) {
    snprintf(msg, 128, "End of buffer reached during a request for %lu bytes (%lu available)", needed, available);
  }
  virtual ~end_of_buffer() throw() {}
  virtual const char* what() const throw() { return msg; }
protected:
  char msg[128];
};

class serialized_buffer {
public:
  serialized_buffer(const char* _p, size_t _len) : p(_p), len(_len) {}

  size_t remaining() const { return len; }
  const char* current() const { return p; }

  const char* get_bytes(size_t bytes) {
    return consume(bytes);
  }

  uint8_t get_u8() {
    return *consume(1);
  }

  uint16_t get_u16_be() {
    return BE16(*reinterpret_cast<const uint16_t*>(consume(2)));
  }
  uint16_t get_u16_le() {
    return LE16(*reinterpret_cast<const uint16_t*>(consume(2)));
  }
  uint32_t get_u24_be() {
    return BE32((*reinterpret_cast<const uint32_t*>(consume(3))) & 0x00ffffff);
  }
  uint32_t get_u24_le() {
    return LE32((*reinterpret_cast<const uint32_t*>(consume(3))) & 0x00ffffff);
  }
  uint32_t get_u32_be() {
    return BE32(*reinterpret_cast<const uint32_t*>(consume(4)));
  }
  uint32_t get_u32_le() {
    return LE32(*reinterpret_cast<const uint32_t*>(consume(4)));
  }
  uint64_t get_u64_be() {
    return BE64(*reinterpret_cast<const uint64_t*>(consume(8)));
  }
  uint64_t get_u64_le() {
    return LE64(*reinterpret_cast<const uint64_t*>(consume(8)));
  }
  double get_double_le() {
    union {
      uint64_t a;
      double b;
    } _d;
    _d.a = get_u64_le();
    return _d.b;
  }
  double get_double_be() {
    union {
      uint64_t a;
      double b;
    } _d;
    _d.a = get_u64_be();
    return _d.b;
  }

protected:
  serialized_buffer(const serialized_buffer& right); // you probably don't want to copy this...

  const char* consume(size_t bytes) {
    if (len < bytes) throw end_of_buffer(bytes, len);

    const char* r = p;
    p += bytes;
    len -= bytes;
    return r;
  }

  const char* p;
  size_t len;
};



AMFData.h
/*
 * AMFData.h
 *
 * Copyright (c) 2007-2008 Dan Weatherford and Facebook, inc.
 * All rights reserved.
 */

#pragma once

#include "common.h"
#include "fout.h"
#include "serialized_buffer.h"
#include <float.h>
#include <math.h>
#include <netinet/in.h>

#ifdef __APPLE__
  #include <architecture/byte_order.h>
  #define hton64(x) OSSwapHostToBigInt64(x)
#else
  #include <byteswap.h>
  #if __BYTE_ORDER == __LITTLE_ENDIAN
    #define hton64(x) bswap_64(x)
  #else
    #define hton64(x) (x)
  #endif
#endif

enum AMFType {
  AMF_TYPE_DOUBLE = 0,
  AMF_TYPE_BOOLEAN = 1,
  AMF_TYPE_STRING = 2,
  AMF_TYPE_OBJECT = 3,
  AMF_TYPE_NULL = 5,
  AMF_TYPE_UNDEFINED = 6,
  AMF_TYPE_MIXED_ARRAY = 8,
  AMF_TYPE_ARRAY = 10,
  AMF_TYPE_DATE = 11,
  AMF_TYPE_UNSUPPORTED = 13,
  AMF_TYPE_UNKNOWN = 255
};

class AMFData {
public:
  virtual ~AMFData() {}
  virtual string asString() const {
    return string("(unknown AMFData type)");
  }

  virtual double asDouble() const {
    return 0.0;
  }

  virtual bool asBool() const {
    return false;
  }

  virtual AMFType typeID() const {
    return AMF_TYPE_UNKNOWN;
  }

  virtual void write(fout& fp) const {
    throw std::runtime_error("AMFData (subtype AMF_TYPE_UNKNOWN): writing this type is meaningless");
  }

  // Factory to produce the right subclass of AMFData object
  static shared_ptr<AMFData> construct(serialized_buffer& buf);

protected:

} ;

class AMFDouble : public AMFData {
public:
  AMFDouble(double _d) : d(_d) {}
  AMFDouble(serialized_buffer& buf) {
    d = buf.get_double_be();
  }
  virtual AMFType typeID() const { return AMF_TYPE_DOUBLE; }
  virtual double asDouble() const { return d; }
  virtual bool asBool() const { return (fabs(d) > FLT_EPSILON); }
  virtual string asString() const {
    char buf[64];
    sprintf(buf, "%f", d);
    return string(buf);
  }
  virtual void write(fout& fp) const {
    union {
      uint64_t a;
      double b;
    } _d;
    _d.b = d;
    fp.putc(AMF_TYPE_DOUBLE);
    fp.write<uint64_t>(hton64(_d.a));
  }

  double d;
} ;

class AMFBoolean : public AMFData {
public:
  AMFBoolean(bool _d) : d(_d) {}
  AMFBoolean(serialized_buffer& buf) {
    d = (buf.get_u8() != 0);
  }

  virtual AMFType typeID() const { return AMF_TYPE_BOOLEAN; }
  virtual double asDouble() const { return d ? 1.0 : 0.0; }
  virtual bool asBool() const { return d; }
  virtual string asString() const {
    return string(d ? "true" : "false");
  }
  virtual void write(fout& fp) const {
    fp.putc(AMF_TYPE_BOOLEAN);
    fp.putc(d ? 0x01 : 0x00);
  }

  bool d;
} ;

class AMFString : public AMFData {
public:
  AMFString(const string& _d) : d(_d) {}
  AMFString() {}
  AMFString(serialized_buffer& buf) {
    uint16_t l = buf.get_u16_be();
    d = string(buf.get_bytes(l), l);
  }

  virtual AMFType typeID() const { return AMF_TYPE_STRING; }
  virtual double asDouble() const { return d.size() ? 1.0 : 0.0; }
  virtual bool asBool() const { return d.size(); }
  virtual string asString() const { return d; }

  virtual void write(fout& fp) const {
    fp.putc(AMF_TYPE_STRING);
    fp.write<uint16_t>(htons(d.size()));
    fp.write(d.data(), d.size());
  }

  string d;
} ;

class AMFNull : public AMFData {
public:
  AMFNull() {}
  AMFNull(serialized_buffer& buf) { }

  virtual AMFType typeID() const { return AMF_TYPE_NULL; }
  virtual double asDouble() const { return 0.0; }
  virtual bool asBool() const { return false; }
  virtual string asString() const { return "NULL"; }
  virtual void write(fout& fp) const {
    fp.putc(AMF_TYPE_NULL);
  }
} ;

class AMFUndefined : public AMFData {
public:
  AMFUndefined() {}
  AMFUndefined(serialized_buffer& buf) { }

  virtual AMFType typeID() const { return AMF_TYPE_UNDEFINED; }
  virtual double asDouble() const { return 0.0; }
  virtual bool asBool() const { return false; }
  virtual string asString() const { return "UNDEFINED"; }
  virtual void write(fout& fp) const {
    fp.putc(AMF_TYPE_UNDEFINED);
  }
} ;

class AMFMixedArray : public AMFData {
public:
  AMFMixedArray() {}
  AMFMixedArray(serialized_buffer& buf) {
    buf.get_u32_be(); //skip the useless nkeys thing that only AMFMixedArray has
                      // (and not AMFObject, which derives from this)
    _construct(buf);
  }

  virtual AMFType typeID() const { return AMF_TYPE_MIXED_ARRAY; }
  virtual double asDouble() const { return dmap.size(); }
  virtual bool asBool() const { return dmap.size(); }
  virtual string asString() const {
    string d("{ \n");
    for (map<string, shared_ptr<AMFData> >::const_iterator dmi = dmap.begin(); dmi != dmap.end(); ++dmi) {
      d = d + string("  ") + dmi->first + string(": ") + dmi->second->asString() + string("\n");
    }
    d += string("}");
    return d;
  }
  // copies the contents of the 'right' array into this one
  // if overwrite is true, right's contents overwrite existing keys in this array
  // if overwrite is false, keys existing in this array will not be copied
  virtual void merge(shared_ptr<AMFData> right, bool overwrite) {
    if (right->typeID() != AMF_TYPE_MIXED_ARRAY) {
      throw std::runtime_error("AMFMixedArray::merge: attempt to merge with something other than a MixedArray");
    }
    AMFMixedArray* r = static_cast<AMFMixedArray*>(&(*right));
    for (map<string, shared_ptr<AMFData> >::const_iterator ri = r->dmap.begin(); ri != r->dmap.end(); ++ri) {
      if (overwrite) {
        dmap[ri->first] = ri->second; // straight assignment operator overwrites existing keys.
      }
      else {
        dmap.insert(*ri); // insert(pair<key, value>) will not overwrite existing keys
      }
    }
  }

  virtual void write(fout& fp) const {
    fp.putc(AMF_TYPE_MIXED_ARRAY);
    fp.write<uint32_t>(htonl(dmap.size())); // mixed arrays have this size thing, but objects don't
    _write(fp);
  }

  map<string, shared_ptr<AMFData> > dmap;
protected:
  void _construct(serialized_buffer& buf) {
    do {
      uint16_t l = 0;
      try {
        l = buf.get_u16_be();
      } catch (const end_of_buffer& e) {
        printf("Error deserializing an array element: %s. Array may be incomplete.\n", e.what());
        return;
      }
      if (l == 0) break; // done
      string k = string(buf.get_bytes(l), l);
      dmap.insert(std::make_pair(k, AMFData::construct(buf)));
    } while (true);
    buf.get_u8(); // eat terminator byte (0x09)
  }
  // writing routines common between this and AMFObject
  void _write(fout& fp) const {
    for (map<string, shared_ptr<AMFData> >::const_iterator dmi = dmap.begin(); dmi != dmap.end(); ++dmi) {
      fp.write<uint16_t>(htons(dmi->first.size()));
      fp.write(dmi->first.data(), dmi->first.size());
      dmi->second->write(fp);
    }
    fp.write<uint16_t>(0);
    fp.putc(0x09); // writeback terminator byte
  }
} ;

class AMFObject : public AMFMixedArray {
public:
  AMFObject() {}
  AMFObject(serialized_buffer& buf) {
    _construct(buf);
  }
  virtual AMFType typeID() const { return AMF_TYPE_OBJECT; }
  virtual void write(fout& fp) const {
    fp.putc(AMF_TYPE_OBJECT);
    _write(fp);
  }
} ;

class AMFArray : public AMFData {
public:
  AMFArray() {}
  AMFArray(serialized_buffer& buf) {
    uint32_t len = buf.get_u32_be();
    for (uint32_t s = 0; s < len; ++s) {
      dmap.push_back(AMFData::construct(buf));
    }
  }

  virtual AMFType typeID() const { return AMF_TYPE_ARRAY; }
  virtual double asDouble() const { return dmap.size(); }
  virtual bool asBool() const { return dmap.size(); }
  virtual string asString() const {
    string d("{ \n");
    for (size_t s = 0; s < dmap.size(); ++s) {
      d = d + string("  ") + dmap[s]->asString() + string("\n");
    }
    d += string("}");
    return d;
  }

  virtual void write(fout& fp) const {
    fp.putc(AMF_TYPE_ARRAY);
    fp.write<uint32_t>(htonl(dmap.size()));
    for (size_t s = 0; s < dmap.size(); ++s) {
      dmap[s]->write(fp);
    }
  }
  vector<shared_ptr<AMFData> > dmap;
} ;

class AMFDate:  public AMFData {
public:
  AMFDate() {
    gettimeofday(&tv, &tz);
  }
  AMFDate(serialized_buffer& buf) {
    // double milliseconds since the epoch
    // int16 TZ offset from UTC
    double s = buf.get_double_be() / 1000.0; // convert from msec
    tv.tv_sec = (time_t) floor(s);
    tv.tv_usec = (time_t) ((s - floor(s)) * 1000000.0);
    tz.tz_minuteswest = buf.get_u16_be();
    tz.tz_dsttime = 0;
  }

  virtual AMFType typeID() const { return AMF_TYPE_DATE; }
  virtual double asDouble() const { return (double)tv.tv_sec + ((double)tv.tv_usec / 1000000.0); }
  virtual bool asBool() const { return (tv.tv_sec != 0); }
  virtual string asString() const {
    char buf[256];
    struct tm tmbuf;
    time_t t = tv.tv_sec - (tz.tz_minuteswest * 60);
    gmtime_r(&t, &tmbuf);
    asctime_r(&tmbuf, buf);
    // drop the newline this thing throws in the buffer
    for (char* c = buf; *c != '\0'; ++c) {
      if (*c == '\n') {
        *c = '\0';
        break;
      }
    }
    return string(buf);
  }

  virtual void write(fout& fp) const {
    fp.putc(AMF_TYPE_DATE);
    // get milliseconds since the epoch
    union {
      uint64_t a;
      double b;
    } _d;
    _d.b = (static_cast<double>(tv.tv_sec) * 1000.0) + (static_cast<double>(tv.tv_usec) / 1000.0);
    fp.write<uint64_t>(hton64(_d.a));
    fp.write<int16_t>(htons(tz.tz_minuteswest));
  }

  struct timeval tv;
  struct timezone tz;
} ;

class AMFUnsupported : public AMFData {
public:
  AMFUnsupported() {}
  AMFUnsupported(serialized_buffer& buf) { }

  virtual AMFType typeID() const { return AMF_TYPE_UNSUPPORTED; }
  virtual double asDouble() const { return 0.0; }
  virtual bool asBool() const { return false; }
  virtual string asString() const { return "UNSUPPORTED"; }
  virtual void write(fout& fp) const {
    fp.putc(AMF_TYPE_UNSUPPORTED);
  }
} ;
